python/prairielearn.py

```diff
@@ -58,6 +69,67 @@ class ElementTestData(QuestionData):
     test_type: Literal["correct", "incorrect", "invalid"]
 
 
+class DisplayType(Enum):
+    INLINE = "inline"
+    BLOCK = "block"
+
+
+def grade_answer_parameterized(
+    data: QuestionData,
+    question_name: str,
+    grade_function: Callable[[Any], Tuple[Union[bool, float], Optional[str]]],
+    weight: int = 1,
+) -> None:
+    """
+    Grade question question_name. grade_function should take in a single parameter
+    (which will be the submitted answer) and return a 2-tuple:
+        - The first element of the 2-tuple should either be:
+            - a boolean indicating whether the question should be marked correct
+            - a partial score between 0 and 1, inclusive
+        - The second element of the 2-tuple should either be:
+            - a string containing feedback
+            - None, if there is no feedback (usually this should only occur if the answer is correct)
+    """
+
+    # Create the data dictionary at first
+    data["partial_scores"][question_name] = {"score": 0.0, "weight": weight}
+
+    if question_name not in data["submitted_answers"]:
+        data["format_errors"][question_name] = "No answer was submitted"
+        return
+
+    submitted_answer = data["submitted_answers"][question_name]
+
+    # Run passed-in grading function
+    result, feedback_content = grade_function(submitted_answer)
+
+    # Try converting partial score
+    if isinstance(result, bool):
+        partial_score = 1.0 if result else 0.0
+    elif isinstance(result, (float, int)):
+        assert 0.0 <= result <= 1.0
+        partial_score = result
+    else:
+        assert_never(result)
+
+    # Set corresponding partial score and feedback
+    data["partial_scores"][question_name]["score"] = partial_score
+
+    if feedback_content:
+        data["partial_scores"][question_name]["feedback"] = feedback_content
+
+
+def determine_score_params(score: float) -> Tuple[str, Union[bool, float]]:
+    """Determine score params taken from data dict"""
```

Comment:
It's not immediately obvious to me what this function does. Why is the second element of the tuple either a boolean or a float?

Consider typing the tuple as `Tuple[Literal["correct", "partial", "incorrect"], Union[bool, float]]` instead to reflect the fact that only specific strings will ever be present in the first element of the tuple. (I just learned about `Literal`: https://mypy.readthedocs.io/en/latest/literal_types.html)

Reply:
I'll add the literal change, that seems very reasonable.

The main purpose of this function is to replace this weird block of code that appears everywhere:

https://github.com/PrairieLearn/PrairieLearn/blob/ae347976c4d2c1946d5417fdc28811565b42b341/elements/pl-multiple-choice/pl-multiple-choice.py#L361-L372

This is just setting the appropriate parameter to make the score badge and percentage render for different elements. Incorrect has to be set to True, since setting it to 0 won't force mustache to render the score, but the mustache templates use the "partial" key to store the percentage of partial credit given. I'll add a note about this in the docstring though.